### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by bakpakin** [janet-lang/janet#79]
Janet could use a multithreading API. The API needs to work with existing Janet code, and be cross platform. It also needs a number of properties:

- It should work with both pthreads and the Windows Thread API
- The programmer should be able to create threads and look at their state.
- The programmer should be able to kill threads.
- Threads should not share Janet memory. This is to simplify the GC and not require locks, mutexes, and atomic primitives
- Sharing data between thread should be done with channels and marshalling data into streams of bytes (which is then unmarshalled on the other side). Internally, this mechanism should use shared memory and locks.
- Threads should be pooled. This could be part of a higher level API in janet.

A preliminary version with just pthreads and no sharing data between threads would be a good start.

Suggested API (most of this would be written in C, but here is the Janet interface)

```clojure
(defn thread/create
  "Create a new thread that will run func in a new thread. Func will be marshalled before
  being duplicated in the new thread. This can be implemented in terms of thread/from-marsh."
  [func &opt channel-size]
  ...)

(defn thread/from-marsh
  "Create a new thread from a marshalled function. Can be useful if you are going to create
  a thread multiple times form the same function."
  [marsh-data &opt channel-size]
  ...)

(defn thread/receive
  "Wait for a message to be sent to this thread. Block until a message is received or the
   timeout expires. Returns the message, or nil if no message received."
  [&opt timeout]
  ...)

(defn thread/send
  "Send a message to a thread. Block until there is space in the message queue or the timeout
  expires. Returns the thread if the message was sent, nil if not."
  [thread-to msg &opt timeout]
  ...)

(defn thread/current
  "Get the current thread running"
  []
  ...)

(defn thread/all
  "Get an array of all running threads."
  []
  ...)

(defn thread/status
  "Get the status of a thread. Return :dead, :running, etc."
  [thread]
  ...)

# Possibly

(defn thread/join
  "Block until all threads have completed. Possibly unnecessary due to thread/receive, but
  definitely convenient."
  [& threads]
  ...)
```

**Comment by andrewchambers** [comment]
Some thoughts:

- The thread mailbox could be a dyn value. It may be less 'magic' to expose the mailbox as an abstract type.

- The timeout api should say what the units are (I know its a draft).

**Comment by bakpakin** [comment]
Timeouts will be in seconds, the exact resolution would be implementation dependent. It's also probably better to allow out of order receives for network transparency, but the initial implementation will almost certainly deliver messages in order. As for the thread mailbox, we need to use the return value for detecting timeouts. The exact interface shouldn't have any effect on locking issues.

**Comment by andrewchambers** [comment]
I mean when you call receive you can supply a pattern to match against to define user priority. That is what erlang does, its just a question as to how fundamental that operation is.

**Comment by bakpakin** [comment]
There is a WIP implantation of threads on the threads-3 branch that deviates a bit from this model (a thread has one mailbox per communicating thread, rather than one thread local mailbox). However, I’ll probably switch to the one thread local mailbox as it makes selecting on multiple mailboxes unnecessary. It should also make the case where multiple worker threads send messages to a supervisor more efficient and simpler. I also think it will encourage better designed programs. Having multiple mailboxes encourages programmers to wait on only one of those mailboxes at a time, which means other mailboxes will block.

**Comment by pepe** [comment]
I have found, that some functions, when sent to a thread (or even used from the thread), are nil. 

Reproduction can be seen at: https://github.com/pepe/shawn/tree/update-as-nil

**Comment by bakpakin** [comment]
Thanks @pepe, this has been fixed in bc2bac8cd340eddf9754212fd002763e22224c4a
